src/main/java/ml/shifu/shifu/core/dtrain/gs/GridSearch.java

```diff
@@ -68,10 +80,63 @@
      */
     private int flattenParamsCount;
 
-    public GridSearch(Map<String, Object> rawParams) {
-        assert rawParams != null;
+    public GridSearch(Map<String, Object> rawParams, List<String> configFileContent) {
         this.rawParams = rawParams;
-        parseParams(this.rawParams);
+        List<Map<String, Object>> gridParams = parseParams(configFileContent);
+        if(gridParams != null && gridParams.size() > 0) {
+            assert this.hasHyperParam(rawParams) == false;
+            Set<String> hyperParams = new HashSet<String>();
+            for(Map<String, Object> map: gridParams) {
+                for(Map.Entry<String, Object> gridEntry: map.entrySet()) {
+                    hyperParams.add(gridEntry.getKey());
+                }
+                for(Map.Entry<String, Object> kvEntry: this.rawParams.entrySet()) {
+                    map.put(kvEntry.getKey(), kvEntry.getValue());
+                }
+            }
+            this.flattenParams = gridParams;
+            this.hyperParamCount = hyperParams.size();
+            this.flattenParamsCount = this.flattenParams.size();
+        } else {
+            // empty file content
+            assert rawParams != null;
+            this.hyperParamCount = 0;
+            parseParams(this.rawParams);
+        }
+    }
+
+    private List<Map<String, Object>> parseParams(List<String> configFileContent) {
+        if(configFileContent != null) {
+            Map<String, MetaItem> metaWarehouse = MetaFactory.getModelConfigMeta();
+            List<Map<String, Object>> gridParams = new ArrayList<Map<String, Object>>();
+            for(String configLine: configFileContent) {
+                Map<String, Object> paramsMap = parseParams(metaWarehouse, configLine);
+                if(paramsMap != null) {
+                    gridParams.add(paramsMap);
+                }
+            }
+            return gridParams;
+        }
+        return null;
+    }
+
+    private Map<String, Object> parseParams(Map<String, MetaItem> metaWarehouse, String configLine) {
```

Comment:
how you can explain such format to users, document?

Reply:
updated the log to add hints